jobspec2/types.config.go

```diff
@@ -0,0 +1,299 @@
+package jobspec2
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/hcl/v2"
+	"github.com/hashicorp/nomad/api"
+	"github.com/zclconf/go-cty/cty"
+)
+
+const (
+	variablesLabel = "variables"
+	variableLabel  = "variable"
+	localsLabel    = "locals"
+	vaultLabel     = "vault"
+	taskLabel      = "task"
+
+	inputVariablesAccessor = "var"
+	localsAccessor         = "local"
+)
+
+type jobConfig struct {
+	JobID string `hcl:",label"`
+	Job   *api.Job
+
+	ParseConfig *ParseConfig
+
+	Vault *api.Vault  `hcl:"vault,block"`
+	Tasks []*api.Task `hcl:"task,block"`
+
+	InputVariables Variables
+	LocalVariables Variables
+
+	LocalBlocks []*LocalBlock
+}
+
+var jobConfigSchema = &hcl.BodySchema{
+	Blocks: []hcl.BlockHeaderSchema{
+		{Type: variablesLabel},
+		{Type: variableLabel, LabelNames: []string{"name"}},
+		{Type: localsLabel},
+		{Type: "job", LabelNames: []string{"name"}},
+	},
+}
+
+func (c *jobConfig) decodeBody(body hcl.Body) hcl.Diagnostics {
+	content, diags := body.Content(jobConfigSchema)
+	if len(diags) != 0 {
+		return diags
+	}
+
+	diags = append(diags, c.decodeInputVariables(content)...)
+	diags = append(diags, c.parseLocalVariables(content)...)
+	diags = append(diags, c.collectInputVariableValues(c.ParseConfig.Envs, c.ParseConfig.parsedVarFiles, toVars(c.ParseConfig.ArgVars))...)
+
+	_, moreDiags := c.InputVariables.Values()
+	diags = append(diags, moreDiags...)
+	_, moreDiags = c.LocalVariables.Values()
+	diags = append(diags, moreDiags...)
+	diags = append(diags, c.evaluateLocalVariables(c.LocalBlocks)...)
+
+	nctx := c.EvalContext()
+
+	diags = append(diags, c.decodeJob(content, nctx)...)
+	return diags
+}
+
+// decodeInputVariables looks in the found blocks for 'variables' and
+// 'variable' blocks. It should be called firsthand so that other blocks can
+// use the variables.
+func (j *jobConfig) decodeInputVariables(content *hcl.BodyContent) hcl.Diagnostics {
```

Comment:
nitpick: we use `(c *jobConfig)` on all the other methods.

Reply:
done in 147c8e12c2858f5aad1b8a0e6dc7baed1654562b